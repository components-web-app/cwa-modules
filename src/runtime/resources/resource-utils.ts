export enum CwaResourceTypes {
  ROUTE = 'ROUTE',
  PAGE = 'PAGE',
  LAYOUT = 'LAYOUT',
  PAGE_DATA = 'PAGE_DATA',
  COMPONENT_GROUP = 'COMPONENT_GROUP',
  COMPONENT_POSITION = 'COMPONENT_POSITION',
  COMPONENT = 'COMPONENT'
}

export interface CwaResource {
  '@id': string
  '@type': string
  publishedResource?: string
  _metadata: {
    persisted: boolean
    publishable?: {
      published: boolean
      publishedAt: string
    }
    [key: string]: any
  }
  [key: string]: any
}

type TypeToPathPrefixMap = {
  [T in CwaResourceTypes]: string;
}

const resourceTypeToIriPrefix: TypeToPathPrefixMap = {
  [CwaResourceTypes.ROUTE]: '/_/routes/',
  [CwaResourceTypes.PAGE]: '/_/pages/',
  [CwaResourceTypes.PAGE_DATA]: '/page_data/',
  [CwaResourceTypes.LAYOUT]: '/_/layouts/',
  [CwaResourceTypes.COMPONENT_GROUP]: '/_/component_groups/',
  [CwaResourceTypes.COMPONENT_POSITION]: '/_/component_positions/',
  [CwaResourceTypes.COMPONENT]: '/component/'
}

export function getResourceTypeFromIri (iri: string): CwaResourceTypes|undefined {
  for (const type of Object.values(CwaResourceTypes)) {
    const prefix: string = resourceTypeToIriPrefix[type]
    if (iri.startsWith(prefix)) {
      return type
    }
  }
}

export function getPublishedResourceIri (resource: CwaResource): string|null {
  const publishableMetadata = resource._metadata?.publishable
  const resourceIri = resource['@id']
  // not a publishable resource
  if (!publishableMetadata) {
    return resourceIri
  }
  if (publishableMetadata.published) {
    return resourceIri
  }
  return resource.publishedResource || null
}

export function isCwaResource (obj: any): obj is CwaResource {
  if (typeof obj !== 'object') {
    return false
  }
  return obj['@id'] !== undefined && obj['@type'] !== undefined && obj._metadata !== undefined && typeof obj._metadata === 'object'
}

export function isCwaResourceSame (resource1: CwaResource, resource2: CwaResource): boolean {
  const clearAndStringify = (obj: CwaResource): string => {
    const newObj: any = Object.assign({}, obj)
    delete newObj.publishedResource
    delete newObj.draftResource
    delete newObj.modifiedAt
    // remove metadata, can include things specific to the resource such as published timestamps
    delete newObj._metadata
    // remove null values
    Object.keys(newObj).forEach(k => newObj[k] === null && delete newObj[k])
    return JSON.stringify(newObj)
  }
  return clearAndStringify(resource1) === clearAndStringify(resource2)
}
