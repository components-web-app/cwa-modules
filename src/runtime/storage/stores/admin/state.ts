import { reactive } from 'vue'

export interface CwaAdminStateInterface {
  state: {
    isEditing: boolean
    navigationGuardDisabled: boolean
  }
}

export default function (): CwaAdminStateInterface {
  return {
    state: reactive({
      isEditing: false,
      navigationGuardDisabled: false,
    }),
  }
}
