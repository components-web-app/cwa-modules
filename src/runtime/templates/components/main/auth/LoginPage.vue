<template>
  <AuthTemplate
    :submitting="submitting"
    :submit-button-text="submitButtonText"
    :error="error"
    @submit="$emit('submit')"
  >
    <template v-if="credentials">
      <InputField
        v-model="credentials.username"
        label="Username or Email"
        name="username"
        type="text"
        autocomplete="username"
        :required="true"
      />
      <InputField
        v-model="credentials.password"
        label="Password"
        name="password"
        type="password"
        autocomplete="current-password"
        :required="true"
      />
    </template>
    <slot>
      <div class="cwa-flex cwa-items-center cwa-justify-between">
        <AuthPageLink link-to="/" link-text="< Back to home" />
        <AuthPageLink link-to="/forgot-password" link-text="Forgot your password?" />
      </div>
    </slot>
  </AuthTemplate>
</template>

<script setup lang="ts">
import { computed, defineAsyncComponent } from 'vue'
import AuthTemplate from './_parts/AuthTemplate.vue'
import InputField from './_parts/InputField.vue'

const emit = defineEmits(['submit', 'update:modelValue'])

const props = defineProps<{
  submitButtonText: string,
  submitting: boolean,
  modelValue: {
    username: string,
    password: string
  }
  error?: string
}>()

const AuthPageLink = defineAsyncComponent(() =>
  import('./_parts/AuthPageLink.vue')
)

const credentials = computed({
  get () {
    return props.modelValue
  },
  set (newValue) {
    emit('update:modelValue', newValue)
  }
})
</script>
