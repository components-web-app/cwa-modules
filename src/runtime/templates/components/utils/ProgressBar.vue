<template>
  <Transition
    enter-from-class="cwa-transform cwa-opacity-0 cwa-w-0"
    enter-active-class="cwa-duration-200 cwa-ease-out"
    enter-to-class="cwa-opacity-100"
    leave-from-class="cwa-opacity-100"
    leave-active-class="cwa-duration-300 cwa-ease-in cwa-delay-200"
    leave-to-class="cwa-transform cwa-opacity-0"
  >
    <div v-show="props.show" class="cwa-w-full cwa-h-1">
      <div class="cwa-bg-green-400 cwa-h-1 cwa-transition-all cwa-duration-300" :style=" { width: `${props.percent}%` }" />
    </div>
  </Transition>
</template>

<script setup>
const props = defineProps({
  show: { required: true, type: Boolean },
  percent: { required: true, type: Number }
})
</script>
