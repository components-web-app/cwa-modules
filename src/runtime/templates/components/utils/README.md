Components in this directory are available to the application automatically with the 'CwaUtils' prefix.

For example:
```vue
<CwaUtilsSpinner :show="true" />
```
