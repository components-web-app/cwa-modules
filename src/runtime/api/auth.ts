import { FetchError } from 'ofetch'
import { computed, ref } from 'vue'
import type { ComputedRef, Ref } from 'vue'
import { useRoute } from '#app'
import type { CookieRef } from '#app'
import { AuthStore } from '../storage/stores/auth/auth-store'
import { CwaUserRoles } from '../storage/stores/auth/state'
import { ResourcesStore } from '../storage/stores/resources/resources-store'
import { FetcherStore } from '../storage/stores/fetcher/fetcher-store'
import CwaFetch from './fetcher/cwa-fetch'
import Mercure from './mercure'
import Fetcher from './fetcher/fetcher'
import Admin from '#cwa/runtime/admin/admin'

interface Credentials {
  username: string
  password: string
}

export enum CwaAuthStatus {
  SIGNED_OUT = 0,
  LOADING = 1,
  SIGNED_IN = 2
}

interface ResetPasswordEvent {
  username: string
  token: string
  passwords: {
    first: string
    second: string
  }
}

export default class Auth {
  private loading: Ref<boolean>
  private hasCheckedMeEndpointForInit = false

  public constructor (
    private readonly cwaFetch: CwaFetch,
    private readonly mercure: Mercure,
    private readonly fetcher: Fetcher,
    private readonly admin: Admin,
    private readonly authStoreDefinition: AuthStore,
    private readonly resourcesStoreDefinition: ResourcesStore,
    private readonly fetcherStoreDefinition: FetcherStore,
    private readonly authCookie: CookieRef<string | null>
  ) {
    this.loading = ref(false)
  }

  public async signIn (credentials: Credentials) {
    const result = await this.loginRequest(credentials)
    if (result instanceof FetchError) {
      return result
    }
    this.mercure.init(true)
    return this.refreshUser()
  }

  public async forgotPassword (username: string) {
    try {
      return await this.cwaFetch.fetch(`/password/reset/request/${encodeURIComponent(
        username
      )}`)
    } catch (error) {
      if (!(error instanceof FetchError)) {
        throw error
      }
      return error
    }
  }

  public async resetPassword (event: ResetPasswordEvent) {
    try {
      return await this.cwaFetch.fetch('/component/forms/password_reset/submit', {
        method: 'PATCH',
        body: {
          password_update: {
            username: event.username,
            plainNewPasswordConfirmationToken: event.token,
            plainPassword: event.passwords
          }
        }
      })
    } catch (error) {
      if (!(error instanceof FetchError)) {
        throw error
      }
      return error
    }
  }

  public async signOut () {
    this.loading.value = true
    try {
      const result = await this.cwaFetch.fetch('/logout')
      await this.clearSession()
      return result
    } catch (error) {
      if (!(error instanceof FetchError)) {
        throw error
      }
      return error
    } finally {
      this.loading.value = false
    }
  }

  public async refreshUser () {
    this.loading.value = true
    try {
      const user = await this.cwaFetch.fetch('/me')
      this.authCookie.value = '1'
      this.authStore.data.user = user
      return user
    } catch (error) {
      if (!(error instanceof FetchError)) {
        throw error
      }
      await this.clearSession()
      return error
    } finally {
      this.loading.value = false
    }
  }

  public async init () {
    if (!this.hasCheckedMeEndpointForInit || (this.signedIn.value && !this.user)) {
      this.hasCheckedMeEndpointForInit = true
      await this.refreshUser()
    }
  }

  public get user () {
    return this.authStore.data.user
  }

  public get roles () {
    if (!this.user) {
      return
    }
    return this.user.roles
  }

  public get signedIn () {
    return computed(() => this.status.value === CwaAuthStatus.SIGNED_IN)
  }

  public hasRole (role: CwaUserRoles|string) {
    if (!this.roles) {
      return false
    }
    return this.roles.includes(role)
  }

  private async loginRequest (credentials: Credentials) {
    this.loading.value = true
    try {
      return await this.cwaFetch.fetch('/login', {
        method: 'POST',
        body: credentials
      })
    } catch (error) {
      this.loading.value = false
      if (!(error instanceof FetchError)) {
        throw error
      }
      return error
    }
  }

  public get status (): ComputedRef<CwaAuthStatus> {
    return computed(() => {
      if (this.loading.value) {
        return CwaAuthStatus.LOADING
      }

      return `${this.authCookie.value}` === '1' ? CwaAuthStatus.SIGNED_IN : CwaAuthStatus.SIGNED_OUT
    })
  }

  private async clearSession () {
    this.authStore.data.user = undefined
    this.authCookie.value = '0'
    this.admin.toggleEdit(false)
    this.mercure.init(true)
    this.resourcesStore.clearResources()
    this.fetcherStore.clearFetches()
    const route = useRoute()
    await this.fetcher.fetchRoute(route)
  }

  private get authStore () {
    return this.authStoreDefinition.useStore()
  }

  private get resourcesStore () {
    return this.resourcesStoreDefinition.useStore()
  }

  private get fetcherStore () {
    return this.fetcherStoreDefinition.useStore()
  }
}
