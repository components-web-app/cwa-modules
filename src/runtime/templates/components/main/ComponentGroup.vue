<template>
  <div v-if="showLoader" class="component-group-placeholder">
    <CwaUtilsSpinner :show="true" />
  </div>
  <CwaUtilsAlertInfo v-else-if="!resource">
    <p>Component Group does not exist, will automatically add when logged in. Functionality coming soon</p>
  </CwaUtilsAlertInfo>
  <template v-else-if="componentPositions && componentPositions.length">
    <ResourceLoader v-for="positionIri of componentPositions" :key="`ResourceLoaderGroupPosition_${resource.value?.data?.['@id']}_${positionIri}`" :iri="positionIri" :ui-component="ComponentPosition" />
  </template>
  <CwaUtilsAlertInfo v-else>
    <p>No component positions in this component group - add functionality coming soon (if logged in)</p>
  </CwaUtilsAlertInfo>
</template>

<script setup>
// todo: draggable drag and drop reordering
// todo: merge in a new component position/ component being added

import { useNuxtApp } from '#app'
import { storeToRefs } from 'pinia'
import { computed, watch } from 'vue'
import _isEqual from 'lodash/isEqual'
import { CwaResourceTypes, getResourceTypeFromIri } from '../../../resources/resource-utils'
import { CwaResourceApiStatuses } from '../../../storage/stores/resources/state'
import ResourceLoader from '../core/ResourceLoader.vue'
import ComponentPosition from '../core/ComponentPosition.vue'
import { CwaAuthStatus } from '../../../api/auth'

const { $cwa } = useNuxtApp()
const resourcesStore = $cwa.storage.stores.resources.useStore()
const { resourcesByType, current: { value: { byId: resources } } } = storeToRefs(resourcesStore)

const props = defineProps({
  reference: { required: true, type: String },
  location: { required: true, type: String },
  allowedComponents: { required: false, type: Array, default () { return null } }
})

const fullReference = computed(() => {
  const locationResource = resources[props.location]
  if (!locationResource) {
    return
  }
  const locationResourceReference = locationResource.data?.reference
  return `${props.reference}_${locationResourceReference}`
})

const resource = computed(() => {
  const componentGroups = resourcesByType.value[CwaResourceTypes.COMPONENT_GROUP]
  return componentGroups.find((componentGroupResource) => {
    return componentGroupResource.data?.reference === fullReference.value
  })
})

const showLoader = computed(() => {
  return $cwa.resources.isLoading.value && (!resource.value || (!resource.value?.data && resource.value?.apiState.status === CwaResourceApiStatuses.IN_PROGRESS))
})

const componentPositions = computed(() => {
  return resource.value?.data?.componentPositions
})

const methods = {
  async createComponentGroup () {
    const resourceTypeProperty = {
      [CwaResourceTypes.PAGE]: 'pages',
      [CwaResourceTypes.LAYOUT]: 'layouts',
      [CwaResourceTypes.COMPONENT]: 'components'
    }
    const locationResourceType = getResourceTypeFromIri(props.location)
    const locationProperty = resourceTypeProperty[locationResourceType]

    const postData = {
      reference: fullReference.value,
      location: props.location,
      allowedComponents: props.allowedComponents
    }
    if (locationProperty) {
      postData[locationProperty] = [props.location]
    }
    await $cwa.resourcesManager.createResource({
      endpoint: '/_/component_groups',
      data: postData
    })
  },
  async updateAllowedComponents () {
    if (_isEqual(props.allowedComponents, resource.value?.data?.allowedComponents)) {
      return
    }
    await $cwa.resourcesManager.updateResource({
      endpoint: resource.value.data['@id'],
      data: {
        allowedComponents: props.allowedComponents
      }
    })
  }
}

watch(() => [$cwa.resources.isLoading.value, $cwa.auth.status.value, resource.value], async ([isLoading, authStatus, resource]) => {
  if (!isLoading && authStatus === CwaAuthStatus.SIGNED_IN) {
    if (!resource) {
      await methods.createComponentGroup()
    } else if (resource.apiState.status === CwaResourceApiStatuses.SUCCESS) {
      await methods.updateAllowedComponents()
    }
  }
}, {
  immediate: true
})
</script>
