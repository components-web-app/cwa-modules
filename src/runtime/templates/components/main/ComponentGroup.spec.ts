// @vitest-environment nuxt
import { describe, expect, test, vi } from 'vitest'
import { shallowMount } from '@vue/test-utils'
import * as nuxt from '#app'
import { nextTick, reactive, ref } from 'vue'
import { mockComponent } from 'vitest-environment-nuxt/utils'
import { CwaResourceTypes } from '@cwa/nuxt3/runtime/resources/resource-utils'
import { CwaAuthStatus } from '@cwa/nuxt3/runtime/api/auth'
import { CwaResourceApiStatuses } from '@cwa/nuxt3/runtime/storage/stores/resources/state'
import ComponentGroup from './ComponentGroup.vue'

const mockReference = 'mockReference'
const mockResourceReference = 'mockResourceReference'
const mockLocation = 'mockLocation'
const mockResourcesByType = { [CwaResourceTypes.COMPONENT_GROUP]: [] }
const mockCwaResources = {
  isLoading: {
    value: false
  }
}
const mockAuth = {
  status: CwaAuthStatus.SIGNED_OUT
}

function createWrapper ({
  reference = mockReference,
  location = mockLocation,
  resourcesByType = mockResourcesByType,
  byId = {},
  cwaResources = mockCwaResources,
  allowedComponents = [],
  auth = mockAuth
}: {
  location?: string;
  resourcesByType?: any;
  reference?: string;
  byId?: any;
  cwaResources?: any;
  allowedComponents?: string[];
  auth?: any;
} = {}) {
  mockComponent('CwaUtilsAlertWarning', () => ({
    name: 'CwaUtilsAlertWarning'
  }))

  mockComponent('CwaUtilsSpinner', () => ({
    name: 'CwaUtilsSpinner',
    props: ['show']
  }))

  mockComponent('CwaUtilsAlertInfo', () => ({
    name: 'CwaUtilsAlertInfo'
  }))

  const mockStore = reactive({
    resourcesByType: ref(resourcesByType),
    current: ref({ byId })
  })
  // @ts-ignore
  vi.spyOn(nuxt, 'useNuxtApp').mockImplementationOnce(() => {
    return {
      $cwa: {
        auth,
        resources: cwaResources,
        resourcesManager: {
          createResource: vi.fn(),
          updateResource: vi.fn()
        },
        storage: {
          stores: {
            resources: {
              useStore () {
                return mockStore
              }
            }
          }
        }
      }
    }
  })

  return shallowMount(ComponentGroup, {
    props: {
      reference,
      location,
      allowedComponents
    },
    global: {
      renderStubDefaultSlot: true
    }
  })
}

describe('ComponentGroup', () => {
  describe('computed properties', () => {
    describe('fullReference', () => {
      test('should return nothing IF there is no resource matching location', () => {
        const wrapper = createWrapper()

        expect(wrapper.vm.fullReference).toBeUndefined()
      })

      test('should return fullReference BASED on location resource data', () => {
        const mockById = {
          [mockLocation]: {
            data: {
              reference: mockResourceReference
            }
          }
        }
        const wrapper = createWrapper({
          byId: mockById
        })

        expect(wrapper.vm.fullReference).toEqual(`${mockReference}_${mockResourceReference}`)
      })
    })

    describe('resource', () => {
      test('should return resource from resources list BASED on type AND full reference', () => {
        const mockGroupElement = { data: { reference: `${mockReference}_${mockResourceReference}` } }
        const mockByType = {
          [CwaResourceTypes.COMPONENT_GROUP]: [mockGroupElement]
        }
        const mockById = {
          [mockLocation]: {
            data: {
              reference: mockResourceReference
            }
          }
        }
        const wrapper = createWrapper({
          resourcesByType: mockByType,
          byId: mockById
        })

        expect(wrapper.vm.resource).toEqual(mockGroupElement)
      })
    })

    describe('showLoader', () => {
      test('should return false IF resources loading flag is false', () => {
        const wrapper = createWrapper({
          cwaResources: {
            isLoading: {
              value: false
            }
          }
        })

        expect(wrapper.vm.showLoader).toEqual(false)
      })

      test('should return true IF resources loading flag is true AND resource is not defined', () => {
        const mockByType = {
          [CwaResourceTypes.COMPONENT_GROUP]: []
        }

        const wrapper = createWrapper({
          resourcesByType: mockByType,
          cwaResources: {
            isLoading: {
              value: true
            }
          }
        })

        expect(wrapper.vm.showLoader).toEqual(true)
      })

      test('should return true IF resources loading flag is true AND resource is in loading state', () => {
        const mockGroupElement = {
          data: {
            reference: `${mockReference}_${mockResourceReference}`
          },
          apiState: {
            status: CwaResourceApiStatuses.IN_PROGRESS
          }
        }
        const mockByType = {
          [CwaResourceTypes.COMPONENT_GROUP]: [mockGroupElement]
        }

        const wrapper = createWrapper({
          resourcesByType: mockByType,
          cwaResources: {
            isLoading: {
              value: true
            }
          }
        })

        expect(wrapper.vm.showLoader).toEqual(true)
      })
    })

    describe('componentPositions', () => {
      test('should return resource component positions BASED on its data', () => {
        const mockComponentPositions = ['pos1', 'pos2', 'pos3']
        const mockGroupElement = {
          data: {
            reference: `${mockReference}_${mockResourceReference}`,
            componentPositions: mockComponentPositions
          }
        }
        const mockByType = {
          [CwaResourceTypes.COMPONENT_GROUP]: [mockGroupElement]
        }
        const mockById = {
          [mockLocation]: {
            data: {
              reference: mockResourceReference
            }
          }
        }
        const wrapper = createWrapper({
          resourcesByType: mockByType,
          byId: mockById
        })

        expect(wrapper.vm.componentPositions).toEqual(mockComponentPositions)
      })
    })
  })

  describe('methods', () => {
    describe('createComponentGroup', () => {
      test('should call resource manager method with correct params', async () => {
        const mockLocation = '/_/pages/mock-page'
        const mockById = {
          [mockLocation]: {
            data: {
              reference: mockResourceReference
            }
          }
        }
        const mockAllowedComponents = ['comp1', 'comp2']

        const wrapper = createWrapper({
          location: mockLocation,
          byId: mockById,
          allowedComponents: mockAllowedComponents
        })

        await wrapper.vm.methods.createComponentGroup()

        expect(wrapper.vm.$cwa.resourcesManager.createResource).toHaveBeenCalledWith({
          data: {
            reference: wrapper.vm.fullReference,
            location: '/_/pages/mock-page',
            allowedComponents: mockAllowedComponents,
            pages: ['/_/pages/mock-page']
          },
          endpoint: '/_/component_groups'
        })
      })

      test('should call resource manager method with correct params without location option IF no such location type is available', async () => {
        const mockLocation = '/_/routes/mock-route'
        const mockById = {
          [mockLocation]: {
            data: {
              reference: mockResourceReference
            }
          }
        }
        const mockAllowedComponents = ['comp1', 'comp2']

        const wrapper = createWrapper({
          location: mockLocation,
          byId: mockById
        })

        await wrapper.setProps({ allowedComponents: mockAllowedComponents })

        await wrapper.vm.methods.createComponentGroup()

        expect(wrapper.vm.$cwa.resourcesManager.createResource).toHaveBeenCalledWith({
          data: {
            reference: wrapper.vm.fullReference,
            location: mockLocation,
            allowedComponents: mockAllowedComponents
          },
          endpoint: '/_/component_groups'
        })
      })
    })

    describe('updateAllowedComponents', () => {
      test('should NOT perform update operation IF allowed components from props and resource are equal', async () => {
        const mockGroupElement = {
          data: {
            reference: `${mockReference}_${mockResourceReference}`,
            allowedComponents: ['comp1']
          }
        }
        const mockByType = {
          [CwaResourceTypes.COMPONENT_GROUP]: [mockGroupElement]
        }
        const mockById = {
          [mockLocation]: {
            data: {
              reference: mockResourceReference
            }
          }
        }
        const wrapper = createWrapper({
          resourcesByType: mockByType,
          byId: mockById,
          allowedComponents: ['comp1']
        })

        await wrapper.vm.methods.updateAllowedComponents()

        expect(wrapper.vm.$cwa.resourcesManager.updateResource).not.toHaveBeenCalled()
      })

      test('should perform update operation IF allowed components from props and resource are NOT equal', async () => {
        const mockId = '/_/test'
        const mockGroupElement = {
          data: {
            reference: `${mockReference}_${mockResourceReference}`,
            allowedComponents: ['comp1'],
            '@id': mockId
          }
        }
        const mockByType = {
          [CwaResourceTypes.COMPONENT_GROUP]: [mockGroupElement]
        }
        const mockById = {
          [mockLocation]: {
            data: {
              reference: mockResourceReference
            }
          }
        }
        const wrapper = createWrapper({
          resourcesByType: mockByType,
          byId: mockById,
          allowedComponents: ['comp2']
        })

        await wrapper.vm.methods.updateAllowedComponents()

        expect(wrapper.vm.$cwa.resourcesManager.updateResource).toHaveBeenCalledWith({
          endpoint: mockId,
          data: {
            allowedComponents: wrapper.props().allowedComponents
          }
        })
      })
    })
  })

  describe('watch', () => {
    test('should NOT create group OR update allowed components IF resources are loading', async () => {
      const mockLoading = ref(false)
      const wrapper = createWrapper({
        cwaResources: {
          isLoading: mockLoading
        }
      })
      const createGroupSpy = vi.spyOn(wrapper.vm.methods, 'createComponentGroup').mockImplementation(() => {})
      const updateAllowedComponentsSpy = vi.spyOn(wrapper.vm.methods, 'updateAllowedComponents').mockImplementation(() => {})

      mockLoading.value = true

      await nextTick()

      expect(createGroupSpy).not.toHaveBeenCalled()
      expect(updateAllowedComponentsSpy).not.toHaveBeenCalled()
    })

    test('should NOT create group OR update allowed components IF user is unauthorized', async () => {
      const mockLoading = ref(true)
      const wrapper = createWrapper({
        cwaResources: {
          isLoading: mockLoading
        },
        auth: {
          status: ref(CwaAuthStatus.SIGNED_OUT)
        }
      })
      const createGroupSpy = vi.spyOn(wrapper.vm.methods, 'createComponentGroup').mockImplementation(() => {})
      const updateAllowedComponentsSpy = vi.spyOn(wrapper.vm.methods, 'updateAllowedComponents').mockImplementation(() => {})

      mockLoading.value = false

      await nextTick()

      expect(createGroupSpy).not.toHaveBeenCalled()
      expect(updateAllowedComponentsSpy).not.toHaveBeenCalled()
    })

    test('should create group IF resources are not loading AND user is authorized AND resource does not exist', async () => {
      const mockStatus = ref(CwaAuthStatus.SIGNED_OUT)
      const wrapper = createWrapper({
        cwaResources: {
          isLoading: ref(false)
        },
        auth: {
          status: mockStatus
        }
      })

      const createGroupSpy = vi.spyOn(wrapper.vm.methods, 'createComponentGroup').mockImplementation(() => {})
      const updateAllowedComponentsSpy = vi.spyOn(wrapper.vm.methods, 'updateAllowedComponents').mockImplementation(() => {})

      mockStatus.value = CwaAuthStatus.SIGNED_IN

      await nextTick()

      expect(createGroupSpy).toHaveBeenCalled()
      expect(updateAllowedComponentsSpy).not.toHaveBeenCalled()
    })

    test('should update allowed components IF resources are not loading AND user is authorized AND resource api status is success', async () => {
      const mockGroupElement = {
        data: {
          reference: `${mockReference}_${mockResourceReference}`
        },
        apiState: {
          status: CwaResourceApiStatuses.SUCCESS
        }
      }
      const mockByType = {
        [CwaResourceTypes.COMPONENT_GROUP]: [mockGroupElement]
      }
      const mockById = {
        [mockLocation]: {
          data: {
            reference: mockResourceReference
          }
        }
      }
      const mockStatus = ref(CwaAuthStatus.SIGNED_OUT)
      const wrapper = createWrapper({
        cwaResources: {
          isLoading: ref(false)
        },
        auth: {
          status: mockStatus
        },
        resourcesByType: mockByType,
        byId: mockById
      })

      const createGroupSpy = vi.spyOn(wrapper.vm.methods, 'createComponentGroup').mockImplementation(() => {})
      const updateAllowedComponentsSpy = vi.spyOn(wrapper.vm.methods, 'updateAllowedComponents').mockImplementation(() => {})

      mockStatus.value = CwaAuthStatus.SIGNED_IN

      await nextTick()

      expect(createGroupSpy).not.toHaveBeenCalled()
      expect(updateAllowedComponentsSpy).toHaveBeenCalled()
    })
  })
})
