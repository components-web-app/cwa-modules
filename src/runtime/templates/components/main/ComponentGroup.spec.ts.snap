// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ComponentGroup > snapshots > should match snapshot IF loader is shown 1`] = `
<div
  class="component-group-placeholder"
>
  <spinner-stub
    show="true"
  />
</div>
`;

exports[`ComponentGroup > snapshots > should match snapshot IF resource is not found 1`] = `<!--v-if-->`;

exports[`ComponentGroup > snapshots > should match snapshot IF there are component positions defined 1`] = `
<div
  data-v-app=""
>
  
  
  <resourceloader-stub
    componentprefix=""
    iri="pos1"
    uicomponent="[object Object]"
  />
  <resourceloader-stub
    componentprefix=""
    iri="pos2"
    uicomponent="[object Object]"
  />
  <resourceloader-stub
    componentprefix=""
    iri="pos3"
    uicomponent="[object Object]"
  />
  
  <!--CWA_MANAGER_END-->
  
</div>
`;

exports[`ComponentGroup > snapshots > should match snapshot IF there are no component positions defined 1`] = `<!--v-if-->`;

exports[`ComponentGroup > snapshots > should match snapshot IF user is logged in but there are no positions 1`] = `
<info-stub>
  <p>
    No component positions in this component group - add functionality coming soon
  </p>
</info-stub>
`;
